lib/srv/alpnproxy/conn.go

```diff
@@ -131,7 +131,7 @@ func (c *PingConn) Read(p []byte) (int, error) {
 
 	err := c.discardPingReads()
 	if err != nil {
```

Comment:
I'm not convinced this properly returns the number of bytes read from the connection.

Looks like we use `binary.Read` under the hood which requires some extra error checking to determine what was read.

Reply:
So diving into the code here:

What this thing does is look for the first non-zero uint32. That uint32 is the size of the next packet, so then it's consciously making the decision to only calc the number of bytes after that size value. Semantically, it's stating that the ping reads are not part of the calc. It just zips past those to the next packet of interest.

So looks like this is properly returning the number of bytes it intends to.